### DISCUSSION TITLE
Basis of spineless fermions in terms of unit cell occupation

### DISCUSSION THREAD

**Discussion by Sreema1996** [QuSpin/QuSpin#611] [General]
I want to implement the particle number conservation in each unit cell as a symmetry of the QuSpin basis. As an example, we can consider the SSH model with $\gamma_{1}$=1, and $\gamma_{2}$=0. For this case, the Hamiltonian reads as 

H=\sum_{j=1} ^{L/2}(c_{2j}^{\dagger}c_{2j+1}+H.C.)

As we can clearly see, the particle number in each unit cell is conserved for this case. Hence, I would like to compute the energy eigenstates in unit cell number basis. Is there any way to write the basis in terms of unit cell occupation? Any help will be appreciated.

**Comment by PFmix** [comment]
Hi,

To the best of my understanding, what you effectively want to do is a post-selection of your basis: to decompose it into sectors for which the Hamiltonian is block-diagonal.

As a recent dive in the documentation says:

> Arbitrary post selection of basis states can be done in QuSpin using the user_basis class, see [A tutorial on QuSpin’s user_basis](https://quspin.github.io/QuSpin/user_basis.html#user-basis-label).

So my advice would be to use `user_basis`... which I do not understand, so I cannot help you more.
If you figure out how to use it, I'll be happy if you could explain it to me.

**Comment by Sreema1996** [comment]
Thank you so much for your reply!

I'm reading the newly added example of the Kitaev model where such user-defined symmetry has been used. For the Kitaev case, the plaquette operator, i.e, the product of sigma matrices, are the conserved quantities. For my case, the sum of two number operators within each unit cell will be conserved. For a system with 4 lattice sites, there will be two such conserved quantities, i.e. $n_{0}+n_{1}$, $n_{2}+n_{3}$. Now, one can use these operators as a symmetry map along with the particle conservation with a set of quantum numbers {0/1/2,0/1/2}. For the Kitaev case, the action of product operators is defined as follows:

def W_symm(state,N,sign_ptr,args):
	""" 
	applies plaquette operator W = "xyzxyz" on sites defined in args. 
	# CAUTION: 32-bit integers code only!

	- the two 'z'-operators in W are easiest to apply, since just return the occupation number in the spin basis.
	- the two 'x'-operators in W are alo easy -- they only flip the occupation
	- the two 'y'-operators in W do both: they flip the occupation and mutiply the state by +/-i (each) 
	"""
	assert(N==8) # works only for N=8!
	out = 0
	#
	# compute sign by applying operators from W on state
	sign_ptr[0] *= -1 + 2*_compute_occupation(state,args[1],N) # 'y'; factor i taken into account below
	sign_ptr[0] *= -1 + 2*_compute_occupation(state,args[2],N) # 'z'
	sign_ptr[0] *= -1 + 2*_compute_occupation(state,args[4],N) # 'y'; factor i taken into account below
	sign_ptr[0] *= -1 + 2*_compute_occupation(state,args[5],N) # 'z'
	sign_ptr[0] *= -1 # -1=i**2, yy 
	#
	# flip occupation of state on sites 0,1, 3,4
	out = _flip_occupation(state,args[0],N) # 'x'
	out = _flip_occupation(out,  args[1],N) # 'y'
	out = _flip_occupation(out,  args[3],N) # 'x'
	out = _flip_occupation(out,  args[4],N) # 'y'
	#
	return out
# the argument list stores the sites for the four W operators, defined clockwise
W_args_1=np.array([0,1,2,3,4,5],dtype=np.uint32) # plaquette 1
W_args_2=np.array([4,3,6,1,0,7],dtype=np.uint32) # plaquette 2
W_args_3=np.array([6,7,0,5,2,1],dtype=np.uint32) # plaquette 3
W_args_4=np.array([2,5,4,7,6,3],dtype=np.uint32) # plaquette 4
#
######  construct user_basis 
#
# define maps dict
q_nums=[0,0,0,0] # not all sectors contain a finite-number of states
maps = dict( W1_block=(W_symm,2,q_nums[0],W_args_1), 
			 W2_block=(W_symm,2,q_nums[1],W_args_2), 
			 W3_block=(W_symm,2,q_nums[2],W_args_3), 
			 W4_block=(W_symm,2,q_nums[3],W_args_4),
			) 
op_dict = dict(op=op,op_args=op_args)




If I want to replace this product with a sum, How should I modify it? Can I do it like this?



def W_symm(state,N,sign_ptr,args):
	""" 
	
	# CAUTION: 32-bit integers code only!


	assert(N==4) # works only for N=4!
	out = 0
	#
	# compute sign by applying operators from W on state
	sign_ptr[0] *= _compute_occupation(state,args[0],N) # 'n'
        sign_ptr[0] *= _compute_occupation(state,args[1],N) # 'n'
	
	return out
# the argument list stores the sites for the four W operators, defined clockwise
W_args_1=np.array([0],dtype=np.uint32)+np.array([1],dtype=np.uint32) # unit cell 1
W_args_2=np.array([2],dtype=np.uint32)+np.array([3],dtype=np.uint32) # unit cell 2

#
######  construct user_basis 
#
# define maps dict
q_nums=[0,0] # not all sectors contain a finite-number of states
maps = dict( W1_block=(W_symm,3,q_nums[0],W_args_1), 
			 W2_block=(W_symm,3,q_nums[1],W_args_2), 
			) 
op_dict = dict(op=op,op_args=op_args)


The main confusing point is how to define W_arg_1 and W_arg_2  when the symmetry operators are the sum of two number operators.

It would be really appreciated if you, please clarify this point.

**Comment by Sreema1996** [comment]
@mgbukov 

Any help regarding the definition of W_arg_1 and W_arg_2 in terms of the sum of occupation operators within one unit cell will be appreciated. Can you please help me to sort out this issue?
